libvast/vast/bitmap_algorithms.hpp

```diff
@@ -466,4 +468,40 @@ auto all(const Bitmap& bm) {
   return !any<!Bit>(bm);
 }
 
+/// Traverses an ID sequence in conjunction with an arbitrary range that
+/// represents blocks of IDs.
+/// @param ids The ID sequence to *select*.
+/// @param begin An iterator to the beginning of the other range.
+/// @param end An iterator to the end of the other range.
+/// @param f A function that transforms *begin* into a half-open interval of IDs
+///          *[x, y)* where *x* is the first and *y* one past the last ID.
+/// @param g A function the performs a user-defined action if the current range
+///          values falls into *(x, y)*, where `(x, y) = f(*begin)` .
+template <class IDs, class Iterator, class F, class G>
+caf::error traverse(const IDs& ids, Iterator begin, Iterator end, F f, G g) {
+  auto rng = select(ids);
+  // TODO: since the table slices have non-decreasing ID offsets, we should use
+  // binary search here to locate the starting point.
+  for (; begin != end; ++begin) {
+    // We're done if there are no more IDs.
+    if (!rng)
+      return caf::none;
+    auto [first, last] = f(*begin);
+    // Make the ID range catch up if it's behind.
+    if (rng.get() < first) {
+      rng.skip(first);
```

Comment:
@mavam are you sure this is correct? The way I read the implementation of `skip` (since it has no documentation...): it skip `n` 1s, not `n` positions. For example, calling `skip(100)` on some `xs` that alternate between 0 and 1 (i.e. 010101...) would end up at position 200 after skipping 100 1s.